In [3]:
# importing the required library 

import numpy as np
import pandas as pd


In [4]:
# loading the dataset to pandas
data=pd.read_csv('Netflix_stock_data.csv')
data

,Date,Close,High,Low,Open,Volume
0,2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000
1,2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000
2,2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000
3,2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000
4,2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000
...,...,...,...,...,...,...
5954,2026-01-22,83.540001,85.099998,82.980003,85.019997,69113300
5955,2026-01-23,86.120003,86.300003,83.279999,83.430000,64520400
5956,2026-01-26,85.699997,86.940002,85.349998,86.919998,40941900
5957,2026-01-27,85.580002,85.599998,83.879997,85.279999,37790800


🔵 Section 1: Data Understanding (Questions 1–8)

Q1. What is the shape of the dataset — how many rows and columns does it have?



In [5]:
data.shape

(5959, 6)

the data have 5959 rows and 6 columns

Q2. What are the column names and what does each column represent? (Date, Open, High, Low, Close, Volume)

In [6]:
data.columns

Index(['Date', 'Close', 'High', 'Low', 'Open', 'Volume'], dtype='object')

above columns of the dataset represent the following things:

Date= date on which the price details where recorded

open= opening price of the stock

close= closing price of the stock 

high= highest price of the stock 

low = lowest price of the stock

volumn= no of stock traded in a day



Q3. What are the data types of each column — are they correct? (Hint: check if Date is string or datetime)

In [7]:
data.dtypes

Date       object
Close     float64
High      float64
Low       float64
Open      float64
Volume      int64
dtype: object

No they are not correct. Date column have object data type(i.e. string). we can not sort or arrange the string. so it is not correct

chage the string data type of the date column into dataTime?

In [8]:
data['Date']=pd.to_datetime(data['Date'])
data.dtypes

Date      datetime64[ns]
Close            float64
High             float64
Low              float64
Open             float64
Volume             int64
dtype: object

Q4. What is the date range of this dataset — from what year to what year does the data span?


In [9]:
min_date=data.min()
print(min_date)


Date      2002-05-23 00:00:00
Close                0.037286
High                 0.041071
Low                  0.034643
Open                 0.037786
Volume                2856000
dtype: object


In [10]:
print(data['Date'])

0      2002-05-23
1      2002-05-24
2      2002-05-28
3      2002-05-29
4      2002-05-30
          ...    
5954   2026-01-22
5955   2026-01-23
5956   2026-01-26
5957   2026-01-27
5958   2026-01-28
Name: Date, Length: 5959, dtype: datetime64[ns]


In [11]:
start_date=data['Date'].min()
end_date=data['Date'].max()
print(f"starting Date = {start_date} and end Date = {end_date}")

starting Date = 2002-05-23 00:00:00 and end Date = 2026-01-28 00:00:00


Q5. How many years of Netflix stock data does this dataset cover?


In [12]:
start_year=data['Date'].dt.year.min()
end_year=data['Date'].dt.year.max()
no_of_years=end_year-start_year
print(no_of_years)

24


this dataset have 24 years of Netflix stock data


Q6. Are there any missing values in any column? If yes, how many?


In [13]:
data.isnull().sum()

Date      0
Close     0
High      0
Low       0
Open      0
Volume    0
dtype: int64

No, there are no missing values in any columns.

Q7. Are there any duplicate rows or duplicate dates in the dataset?


In [14]:
print(data.duplicated().sum())

0


So there are 0 duplicate rows in the dataset

Q8. What are the minimum, maximum, mean, and median values of the Close price column?


In [15]:
data['Close'].describe()

count    5959.000000
mean       19.793527
std        27.840294
min         0.037286
25%         0.440571
50%         5.074286
75%        33.754499
max       133.912994
Name: Close, dtype: float64

minimum Close price = 0.037286

maximum close price = 133.912994

mean or average close price = 19.793527

median close price = 5.074286

🟡 Section 2: Data Manipulation (Questions 9–17)

Q9. Convert the Date column from string type to datetime format using pd.to_datetime() — then set it as the index.

In [16]:
data['Date']=pd.to_datetime(data['Date']) # change the Date column into dateTime datatype
print(data['Date'].dtype)
print()
data.set_index('Date', inplace=True)
data.head()

datetime64[ns]



,Close,High,Low,Open,Volume
Date,,,,,
2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000
2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000
2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000
2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000
2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000


Q10. Extract new columns from Date — create separate columns for Year, Month, Day, and Day of Week.


In [17]:
data['Year']=data.index.year
data['Month']=data.index.month
data['Day']=data.index.day
data['Day of Week']= data.index.day_name()
data.head()

,Close,High,Low,Open,Volume,Year,Month,Day,Day of Week
Date,,,,,,,,,
2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000,2002,5,23,Thursday
2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000,2002,5,24,Friday
2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000,2002,5,28,Tuesday
2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000,2002,5,29,Wednesday
2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000,2002,5,30,Thursday


Q11. Create a new column called Daily_Range which is the difference between the High and Low price each day — which day had the highest price swing?


In [37]:

data['Daily_Range']=data['High']-data['Low']
data.head()


,Close,High,Low,Open,Volume,Year,Month,Day,Day of Week,Daily_Range
Date,,,,,,,,,,
2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000,2002,5,23,Thursday,0.009715
2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000,2002,5,24,Friday,0.002786
2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000,2002,5,28,Tuesday,0.007500
2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000,2002,5,29,Wednesday,0.007858
2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000,2002,5,30,Thursday,0.003643


Daily_Range column has the range of each day. (i.e. difference between High and Low of the day)

In [43]:
max_daily_range=data['Daily_Range'].max()
max_daily_range_day=data['Daily_Range'].idxmax().day_name()
print(f"{max_daily_range_day} has the highest price swing of {max_daily_range}")

Wednesday has the highest price swing of 9.702995300292969


In [44]:
min_daily_range=data['Daily_Range'].min()
min_daily_range_day=data['Daily_Range'].idxmin().day_name()
print(f"{min_daily_range_day} has the minimum price swing of {min_daily_range}")

Thursday has the minimum price swing of 0.0006420016288757047


Q12. Create a new column called Daily_Return which calculates the percentage change in Close price from one day to the next using .pct_change().

In [52]:
data['Daily_Return']=data['Close'].pct_change()*100
data.head()

,Close,High,Low,Open,Volume,Year,Month,Day,Day of Week,Daily_Range,Daily_Return
Date,,,,,,,,,,,
2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000,2002,5,23,Thursday,0.009715,NaN
2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000,2002,5,24,Friday,0.002786,1.134205
2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000,2002,5,28,Tuesday,0.007500,-4.368596
2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000,2002,5,29,Wednesday,0.007858,-4.629515
2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000,2002,5,30,Thursday,0.003643,-2.912368


Daily_Return column indicates the percentage return on each day

Q13. Filter the data to show only rows from the year 2020 — how did Netflix stock behave during COVID?


In [60]:
data_2020=data[data['Year']== 2020]
data_2020.describe()

,Close,High,Low,Open,Volume,Year,Month,Day,Daily_Range,Daily_Return
count,253.000000,253.000000,253.000000,253.000000,2.530000e+02,253.0,253.000000,253.000000,253.000000,253.000000
mean,44.682723,45.432004,43.836893,44.614146,6.933901e+07,2020.0,6.545455,15.822134,1.595111,0.245365
std,6.561263,6.607927,6.500556,6.606396,3.973960e+07,0.0,3.447294,8.830637,0.791287,2.910743
min,29.884001,32.290001,29.025000,30.240000,1.144000e+07,2020.0,1.000000,1.000000,0.419003,-11.138865
25%,38.105000,38.953999,37.588001,38.000000,4.485800e+07,2020.0,4.000000,8.000000,1.058998,-1.338544
50%,46.995998,47.840000,46.327000,46.854000,5.840500e+07,2020.0,7.000000,16.000000,1.412003,0.141324
75%,49.598999,50.466000,48.825001,49.543999,7.881100e+07,2020.0,10.000000,23.000000,1.955997,1.777214
max,55.654999,57.536999,54.099998,56.798000,2.499140e+08,2020.0,12.000000,31.000000,5.695999,11.608708


Q14. Find the top 10 days where Netflix had the highest trading Volume — what were the dates?


In [69]:
highest_trading_day=data.sort_values('Volume',ascending=False).head(10)
highest_trading_day

,Close,High,Low,Open,Volume,Year,Month,Day,Day of Week,Daily_Range,Daily_Return
Date,,,,,,,,,,,
2004-10-15,0.147143,0.164143,0.142286,0.143286,3234140000,2004,10,15,Friday,0.021857,-40.906423
2011-10-25,1.105286,1.134143,1.060714,1.070000,3155418000,2011,10,25,Tuesday,0.073429,-34.895629
2011-09-20,1.857571,2.042571,1.848143,2.020000,2243430000,2011,9,20,Tuesday,0.194428,-9.544349
2012-01-09,1.402571,1.423571,1.244857,1.280143,2140733000,2012,1,9,Monday,0.178714,13.779100
2013-01-25,2.422286,2.466857,2.080143,2.081000,1914458000,2013,1,25,Friday,0.386714,15.456910
2004-07-16,0.328857,0.387143,0.326571,0.383429,1838361000,2004,7,16,Friday,0.060572,-28.062554
2013-10-22,4.607429,5.559429,4.592857,5.540571,1810998000,2013,10,22,Tuesday,0.966572,-9.146736
2012-07-25,0.861143,0.954286,0.845714,0.917286,1734705000,2012,7,25,Wednesday,0.108572,-25.015568
2004-01-22,0.554714,0.565714,0.523000,0.541357,1662346000,2004,1,22,Thursday,0.042714,18.024261


these are the data of top 10 day with highest volume


In [72]:
highest_trading_day.index

DatetimeIndex(['2004-10-15', '2011-10-25', '2011-09-20', '2012-01-09',
               '2013-01-25', '2004-07-16', '2013-10-22', '2012-07-25',
               '2004-01-22', '2012-01-26'],
              dtype='datetime64[ns]', name='Date', freq=None)

these are the dates of top 10 days with highest volume


Q15. Calculate the average monthly Close price for each year — group by Year and Month.

In [80]:
monthly_averages = data.groupby(['Year', 'Month'])['Close'].mean()
monthly_averages.head(20)


Year  Month
2002  5        0.113583
      6        0.103807
      7        0.111000
      8        0.087097
      9        0.082807
      10       0.056022
      11       0.068418
      12       0.079231
2003  1        0.085112
      2        0.099244
      3        0.127313
      4        0.150612
      5        0.163704
      6        0.163517
      7        0.187682
      8        0.190218
      9        0.251201
      10       0.349441
      11       0.351361
      12       0.361523
Name: Close, dtype: float64

Q16. Create a new column called Price_Direction — label each row as "Up" if Close > Open, and "Down" if Close < Open.


In [82]:
data['Price_Direction']=np.where(data['Close']>data['Open'],"Up","Down")
data

,Close,High,Low,Open,Volume,Year,Month,Day,Day of Week,Daily_Range,Daily_Return,Price_Direction
Date,,,,,,,,,,,,
2002-05-23,0.119643,0.124286,0.114571,0.115643,1047900000,2002,5,23,Thursday,0.009715,NaN,Up
2002-05-24,0.121000,0.122500,0.119714,0.121429,111048000,2002,5,24,Friday,0.002786,1.134205,Down
2002-05-28,0.115714,0.123214,0.115714,0.121357,66094000,2002,5,28,Tuesday,0.007500,-4.368596,Down
2002-05-29,0.110357,0.116429,0.108571,0.116429,67578000,2002,5,29,Wednesday,0.007858,-4.629515,Down
2002-05-30,0.107143,0.110786,0.107143,0.110786,101542000,2002,5,30,Thursday,0.003643,-2.912368,Down
...,...,...,...,...,...,...,...,...,...,...,...,...
2026-01-22,83.540001,85.099998,82.980003,85.019997,69113300,2026,1,22,Thursday,2.119995,-2.132146,Down
2026-01-23,86.120003,86.300003,83.279999,83.430000,64520400,2026,1,23,Friday,3.020004,3.088343,Up
2026-01-26,85.699997,86.940002,85.349998,86.919998,40941900,2026,1,26,Monday,1.590004,-0.487698,Down


Q17. Calculate a 30-day and 90-day rolling average of the Close price and add them as new columns — MA30 and MA90.

In [83]:
# 1. Calculate the 30-day rolling average
data['MA30'] = data['Close'].rolling(window=30).mean()

# 2. Calculate the 90-day rolling average
data['MA90'] = data['Close'].rolling(window=90).mean()

# View the original price alongside the new moving averages
print(data[['Close', 'MA30', 'MA90']].tail())

                Close       MA30        MA90
Date                                        
2026-01-22  83.540001  91.702000  106.646245
2026-01-23  86.120003  91.349001  106.267289
2026-01-26  85.699997  91.115334  105.885611
2026-01-27  85.580002  90.831667  105.471500
2026-01-28  84.639999  90.480001  105.069967
